#16015: Support new op ttnn.experimental.broadcast_to

dchenTT opened the pull request:
### Ticket
Link to Github Issue (https://github.com/tenstorrent/tt-metal/issues/16015)

### Problem description
Add broadcast_to op that is the counter part of torch.broadcast_to. However, currently metal framework does not support tensor strided view, so tensor is allocated.

### What's changed
The broadcast logic is adopted from eltwise binary op broadcast logic, using subtile broadcast. All new files for this PR since it is a new op.

Currently it is an experimental op with basic support as following

      Data Type
	      Bfloat16
	      Float32

      Tensor Shape
	      1D to 4D
      
      Memory Layout
	      Tile
      
      Memory Config
	      Interleaved (DRAM / L1)

Performance Profiling

Before this broadcast_to op, ttnn.repeat can be used as workaround. However, from profiling result, ttnn.repeat can be much slower than broadcast_to, at least for large tensor. Following is the profiling result

| Input                    | Output                   | Subtile broadcast type | Dimension to broadcast |
| ---------------- | ------------------ | ---------------------- | ---------------------- |
| (1, 1, 256, 256) | (4, 128, 256, 256)   | no change                     | N C                    |
| (4, 128, 1, 1)      | (4, 128, 256, 256)   | scalar                             | H W                   |
| (4, 1, 256, 1)     | (4, 128, 256, 256)   | col                                  | C W                   |
| (4, 1, 1, 256)     | (4, 128, 256, 256)   | row                                 | C H                    |

| Subtile broadcast type                                     | No change | Scalar | Col   | Row   |
| ---------------------------------------------------------- | --------- | ------ | ----- | ----- |
| Repeat kernel / broadcast_to kernel duration (in slowness) | 3.27x     | 3.73x  | 2.48x | 1.99x |


### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

First draft

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Use correct way to get output tensor.

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

More gtest cases

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Fix op names

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Added 4 types of reader and writer kernels

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

kernel bug fixes

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Kernels clean up

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Make kernel code more readable

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Rebase and compile

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Use subtile broadcast LLK

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Support pybind with queue_id

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Refactor and clean up

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Rebase again and fix a rebase error

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Fix gtest after rebase

ntarafdar commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/device/kernels/compute/compute_interleaved_col_bcast_to.cpp:
```diff
@@ -0,0 +1,65 @@
+    uint32_t tiles_per_batch = HtWt * C;
+    uint32_t start_n = start_tile_id / tiles_per_batch;
+    uint32_t start_remaining = start_tile_id % tiles_per_batch;
+    uint32_t start_c = start_remaining / HtWt;
```
no division, or modulo in kernels , do it on host and send it in

ntarafdar commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/device/kernels/compute/compute_interleaved_col_bcast_to.cpp:
```diff
@@ -0,0 +1,65 @@
+    uint32_t start_n = start_tile_id / tiles_per_batch;
+    uint32_t start_remaining = start_tile_id % tiles_per_batch;
+    uint32_t start_c = start_remaining / HtWt;
+    uint32_t start_t = start_remaining % HtWt;
```
same as above

ntarafdar commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/device/kernels/compute/compute_interleaved_col_bcast_to.cpp:
```diff
@@ -0,0 +1,65 @@
+    uint32_t start_c = start_remaining / HtWt;
+    uint32_t start_t = start_remaining % HtWt;
+    uint32_t start_th = start_t / Wt;
+    uint32_t start_tw = start_t % Wt;
```
same as above

ntarafdar commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/device/kernels/dataflow/reader_interleaved_col_bcast_to.cpp:
```diff
@@ -0,0 +1,64 @@
+    uint32_t start_c = start_remaining / HtWt;
+    uint32_t start_t = start_remaining % HtWt;
+    uint32_t start_th = start_t / Wt;
+    uint32_t start_tw = start_t % Wt;
```
same feedback , don't put modules or divisions in kernel

dchenTT replied:
Done. And other places.

ntarafdar commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/device/kernels/dataflow/reader_interleaved_col_bcast_to.cpp:
```diff
@@ -0,0 +1,64 @@
+                cb_reserve_back(cb_id_src, onetile);
+                uint32_t l1_write_addr = get_write_ptr(cb_id_src);
+                noc_async_read_tile(tile_offset + th, src, l1_write_addr);
+                noc_async_read_barrier();
```
were can move this barrier, you're not doing a read on the l1_write_addr in the kernel, I think you can move this outside the loop

ntarafdar commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/device/kernels/dataflow/reader_interleaved_no_bcast_to.cpp:
```diff
@@ -0,0 +1,60 @@
+    uint32_t start_n = start_tile_id / tiles_per_batch;
+    uint32_t start_remaining = start_tile_id % tiles_per_batch;
+    uint32_t start_c = start_remaining / HtWt;
+    uint32_t start_t = start_remaining % HtWt;
```
same feedback as others, no modulo or division in kernel

ntarafdar commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/device/kernels/dataflow/reader_interleaved_no_bcast_to.cpp:
```diff
@@ -0,0 +1,60 @@
+                cb_reserve_back(cb_id_src, onetile);
+                uint32_t l1_write_addr_src = get_write_ptr(cb_id_src);
+                noc_async_read_tile(tile_offset++, src, l1_write_addr_src);
+                noc_async_read_barrier();
```
move barrier

ntarafdar commented on tests/ttnn/unit_tests/operations/eltwise/test_broadcast_to.py:
```diff
@@ -0,0 +1,216 @@
+
+
+profile_input_bcast_shape_pairs = [
+    # ((1, 1, 32, 32), (20, 40, 32, 32)),
```
delete commented out code, or put it back

ntarafdar reviewed (changes requested):
some kernel changes

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/bcast_to.cpp:
```diff
@@ -0,0 +1,65 @@
+
+namespace ttnn::operations::experimental {
+
+auto check_shape(const Tensor& input, const ttnn::Shape& output_shape) {
```
best to put into an anonymous namespace.
also because we often build as UNITY build, multiple translation units can get merged together and to avoid symbols clash, we put such things into `CMAKE_UNIQUE_NAMESPACE` namespace, so that cmake ensures that such namespace will truly have a unique name in each translation unit before merging them all together.

ayerofieiev-tt commented on tests/ttnn/unit_tests/operations/eltwise/test_broadcast_to.py:
```diff
@@ -0,0 +1,216 @@
+
+
+input_bcast_shape_pairs = [
+    # ((1, 1, 1, 1), (1, 1, 1)),
+    # ((1, 1, 1, 1), (1, 1, 1, 1, 1)),
+    # ((1, 1, 1, 1, 1), (1, 1, 1, 1, 1)),
+    # ((2,1), (3,1)),
```
remove, uncomment?

ayerofieiev-tt commented on tests/ttnn/unit_tests/operations/eltwise/test_broadcast_to.py:
```diff
@@ -0,0 +1,216 @@
+    torch_result = torch_input_tensor.broadcast_to(broadcast_shape)
+
+    input_tensor = ttnn.from_torch(
+        torch_input_tensor, layout=ttnn.TILE_LAYOUT, device=device, memory_config=memory_config_input
```
does the op work in Row Major?

ayerofieiev-tt commented on tests/ttnn/unit_tests/gtests/test_broadcast_to.cpp:
you can do something like this, just fyi
https://gist.github.com/ayerofieiev-tt/f5305162cbaa358cc3b76d8fe24f82f1

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/device/bcast_to_device_operation.cpp:
```diff
@@ -0,0 +1,114 @@
+#include "ttnn/tensor/types.hpp"
+
+namespace ttnn::operations::experimental::broadcast_to {
+SubtileBroadcastType get_subtile_broadcast_type(uint32_t a_h, uint32_t a_w, uint32_t b_h, uint32_t b_w) {
```
Can you please check 2 things?

this does not seem to be handled `a is (1,1) and b is (1,5)`


`a is (1,1) and b is (5,1)`, both the SCALAR and ROW conditions could conceptually apply, but the ROW case won't be reached because the SCALAR check comes first. Is that okay?

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/device/bcast_to_device_operation.cpp:
```diff
@@ -0,0 +1,114 @@
+
+    switch (input.get_layout()) {
+        case Layout::TILE: return BcastToTileFactory{};
+        default: TT_THROW("BcastTo: Unsupported input layout");
```
https://github.com/tenstorrent/tt-metal/blob/main/contributing/ErrorMessageBestPractices.md

```
TT_FATAL(input.get_layout() == Layout::TILE", BcastTo operation failed: Input layout {} is not supported. Currently, only TILE layout is supported. Please convert your input tensor to TILE layout before performing this operation.", input.get_layout());
```

ayerofieiev-tt replied:
I am also curios why we only support tile layout

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/device/bcast_to_program_factory.cpp:
```diff
@@ -0,0 +1,216 @@
+        tt::tt_metal::SetRuntimeArgs(program, kernel_id, core, args);
+    };
+
+    set_or_update_runtime_arguments(
```
wonderful!

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/device/kernels/compute/compute_interleaved_row_bcast_to.cpp:
```diff
@@ -0,0 +1,67 @@
+            for (uint32_t th = start_th; th < Ht && num_tiles_read < num_tiles; ++th, start_tw = 0) {
+                for (uint32_t tw = start_tw; tw < Wt && num_tiles_read < num_tiles; ++tw) {
+                    cb_wait_front(tt::CBIndex::c_0, 1);
+                    // DeviceZoneScopedN("bcast_row_compute");
```
please remove commented code

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/device/kernels/compute/compute_interleaved_scalar_bcast_to.cpp:
```diff
@@ -0,0 +1,64 @@
+    uint32_t start_th = start_t / Wt;
+    uint32_t start_tw = start_t % Wt;
+
+    // this is the INPUT tile offset
```
please remove commented code

ayerofieiev-tt reviewed (commented):
no input on kernels. 
otherwise looks very decent!
please ping me for the final approve. thanks!

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Addressed code review comments

dchenTT commented:
Hi Guys, could you review the latest code again? Thanks!

patrickroberts commented on tests/ttnn/unit_tests/gtests/test_broadcast_to.cpp:
```diff
@@ -0,0 +1,89 @@
+    auto& device = *device_;
+    std::array<uint32_t, 4> dimensions = {param.n, param.c, param.h, param.w};
+    ttnn::Shape input_shape(dimensions);
+    std::vector<uint32_t> output_size = {
+        param.broadcast_shape[0],
+        param.broadcast_shape[1],
+        param.broadcast_shape[2],
+        param.broadcast_shape[3],
+    };
+    ttnn::Shape output_shape(output_size);
```
If you make `BroadcastParam::broadcast_shape` a `std::vector<uint32_t>` then you could just...
```suggestion
    ttnn::Shape output_shape(param.broadcast_shape);
```

patrickroberts commented on tests/ttnn/unit_tests/gtests/test_broadcast_to.cpp:
```diff
@@ -0,0 +1,89 @@
+    };
+    ttnn::Shape output_shape(output_size);
+
+    const std::optional<MemoryConfig> memory_config = std::nullopt;
+    const std::optional<ttnn::Tensor> output = std::nullopt;
```
These can also be `auto`. The parameters below would just be deduced as `std::nullopt_t` which still works fine.

patrickroberts commented on tests/ttnn/unit_tests/operations/eltwise/test_broadcast_to.py:
```diff
@@ -0,0 +1,216 @@
+from tests.tt_eager.python_api_testing.sweep_tests.generation_funcs import gen_func_with_cast_tt
+
+input_bcast_shape_pairs = [
+    ((1), (2, 10)),
+    ((1, 1, 1, 1), (1, 1, 1, 67000)),
+    ((1, 1, 1, 1), (1, 1, 67000, 4)),
+    ((1, 1, 1, 1), (1, 1, 270, 270)),
+    ((1, 1, 270, 270), (1, 1, 270, 270)),
+    ((1, 1, 64, 64), (1, 310, 64, 64)),
+    ((1, 1, 64, 64), (310, 1, 64, 64)),
+    ((1, 1, 64, 64), (1, 310, 64, 64)),
+    ((31, 33, 64, 64), (31, 33, 64, 64)),
+]
```
Be careful, `(x)` is just `x`, but `(x,)` is a tuple of `x`.

patrickroberts commented on tests/ttnn/unit_tests/operations/eltwise/test_broadcast_to.py:
```diff
@@ -0,0 +1,216 @@
+    ((2, 16, 1, 1), (2, 16, 64, 64)),
+    ((2, 1, 64, 1), (2, 16, 64, 64)),  # col
+    ((2, 1, 1, 64), (2, 16, 64, 64)),  # row
+    # ((1, 1, 256, 256), (4, 128, 256, 256)),
+    # ((4, 128, 1, 1), (4, 128, 256, 256)),
+    # ((4, 1, 256, 1), (4, 128, 256, 256)), # col
+    # ((4, 1, 1, 256), (4, 128, 256, 256)), # row
```
Are these failing?

dchenTT replied:
No. It takes longer time to run because tensor is large. Now I added a comment saying why they are left there.

patrickroberts commented on tests/ttnn/unit_tests/operations/eltwise/test_broadcast_to.py:
```diff
@@ -0,0 +1,216 @@
+        _ = ttnn.experimental.broadcast_to(a_tt, bcast)
+
+
+input_bcast_shape_pairs = [
+    ((1, 1), (7, 10, 32, 32)),
+]
+
+
+@pytest.mark.parametrize("dtype_pt, dtype_tt", ((torch.bfloat16, ttnn.bfloat16),))
+@pytest.mark.parametrize("input, bcast", input_bcast_shape_pairs)
```
If this is only used once, I think it's more readable to avoid naming this since the parameter names in `pytest.mark.parameterize` are already sufficient to identify their purpose, keeps things more local.

dchenTT replied:
moved inside, since it is currently not parameterized.

patrickroberts commented on tests/ttnn/unit_tests/operations/eltwise/test_broadcast_to.py:
```diff
@@ -0,0 +1,216 @@
+        layout=ttnn.TILE_LAYOUT,
+    )
+
+    with pytest.raises(RuntimeError):
+        _ = ttnn.experimental.broadcast_to(a_tt, bcast)
```
Suggestion but not request: Do we want to test for a specific error message?

dchenTT replied:
We can. This is the minimum code to catch runtime error, assuming it does not hit other run time error coincidently.

patrickroberts commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/bcast_to.cpp:
```diff
@@ -0,0 +1,68 @@
+namespace CMAKE_UNIQUE_NAMESPACE {
+auto check_shape(const Tensor& input, const ttnn::Shape& output_shape) {
+    auto input_shape = input.get_logical_shape();
+    TT_FATAL(
+        input_shape.size() <= output_shape.size(),
+        "Input tensor shape {}({}) must be at least as large as the broadcast shape {}({}), which it is not",
+        input_shape,
+        input_shape.size(),
+        output_shape,
+        output_shape.size());
+
+    TT_FATAL(
+        input_shape.size() <= 4 and output_shape.size() <= 4,
+        "Tensor shape and broadcast size {}({}) {}({}) must be at most 4D",
+        input_shape,
+        input_shape.size(),
+        output_shape,
+        output_shape.size());
```
Nice error messages!

patrickroberts commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/bcast_to.cpp:
```diff
@@ -0,0 +1,68 @@
+    size_t input_ndim = input_shape.size();
+    size_t output_ndim = output_shape.size();
+
+    for (int i = -1; i >= -input_ndim; --i) {
```
`i >= -input_ndim` will never be `true` here: https://godbolt.org/z/M9KqnrbYd

dchenTT replied:
added static_cast.

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Addressed code review comments.

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Add test case for bfloat8_b

ayerofieiev-tt commented on tests/ttnn/unit_tests/gtests/test_broadcast_to.cpp:
```diff
@@ -0,0 +1,83 @@
+namespace test {
+
+struct BroadcastParam {
+    uint32_t n;  // input tensor batch
+    uint32_t c;  // input tensor channel
+    uint32_t h;  // input tensor height
+    uint32_t w;  // input tensor width
```
```suggestion
    uint32_t n = 0;  // input tensor batch
    uint32_t c = 0;  // input tensor channel
    uint32_t h = 0;  // input tensor height
    uint32_t w = 0;  // input tensor width
```

ayerofieiev-tt replied:
https://github.com/tenstorrent/tt-metal/blob/main/contributing/BestPractices.md#15-initialize-primitive-types-on-declaration

ayerofieiev-tt commented on tests/ttnn/unit_tests/gtests/test_broadcast_to.cpp:
```diff
@@ -0,0 +1,83 @@
+    ChannelAndBatch,
+    Broadcast_toFixture,
+    ::testing::Values(
+        BroadcastParam{1, 1, 64, 64, {1, 3, 64, 64}},  // channel expansion
+        BroadcastParam{1, 1, 64, 64, {3, 1, 64, 64}}   // batch expansion
```
can we add a test for both?
can we add a test for more combined expansions?

ayerofieiev-tt replied:
can you please also add c++ tests for non-tile aligned dimensions?

dchenTT replied:
when I first develop the op, I used gtest rather than pytest, because it is easier to run gdb debug. It only covers core cases. Once the op is stable, further tests are moved to pytest. gtest cases remain basic. I will add more cases in next commit.

ayerofieiev-tt commented on tests/ttnn/unit_tests/gtests/test_broadcast_to.cpp:
really appreciate a C++ test here.
thank you

ayerofieiev-tt commented on tests/ttnn/unit_tests/operations/eltwise/test_broadcast_to.py:
```diff
@@ -0,0 +1,246 @@
+    input_tensor = ttnn.from_torch(
+        torch_input_tensor, layout=ttnn.TILE_LAYOUT, device=device, memory_config=memory_config_input
+    )
+    output = ttnn.experimental.broadcast_to(
```
how do you see a path toward moving this from experimental?

ayerofieiev-tt reviewed (commented):
@dchenTT can you please also add a test for program cache to ensure that 
- program is reused as expected
- new program is created as expected

You can check number of entries in program cache to make sure that its as spected.

ayerofieiev-tt reviewed (commented):
I might have missed it, but if it does not exist, please also add a test checking that operation works as expected when preallocated output tensor is provided.

ayerofieiev-tt reviewed (approved):
approve so you don't have to wait for me again once feedback is addressed.
Thank you!

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Add more test cases to gtest

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Fixed a unity build error in ccl, also fixed a non-unity build error in bcast_to.

jvegaTT commented on tests/ttnn/unit_tests/gtests/test_broadcast_to.cpp:
```diff
@@ -0,0 +1,110 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
```
2025

dchenTT replied:
OK. Double checked other files, this is the only one.

jvegaTT commented on tests/ttnn/unit_tests/gtests/test_broadcast_to.cpp:
```diff
@@ -0,0 +1,110 @@
+namespace experimental {
+namespace broadcast_to {
+namespace test {
+
```
You can combine these into a single line to make it more readable

dchenTT replied:
Sure.

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/device/kernels/compute/compute_interleaved_col_bcast_to.cpp:
```diff
@@ -0,0 +1,56 @@
+    uint32_t Ht = get_arg_val<uint32_t>(10);
+    uint32_t Wt = get_arg_val<uint32_t>(11);
+
+    unary_bcast_init<BroadcastType::COL>(tt::CBIndex::c_0, tt::CBIndex::c_16);
```
Pass this via CT

dchenTT replied:
Many people did not do that, I will follow higher standard. :-)

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/device/kernels/compute/compute_interleaved_col_bcast_to.cpp:
```diff
@@ -0,0 +1,56 @@
+    for (uint32_t n = start_n; n < N && num_tiles_read < num_tiles; ++n, start_c = 0) {
+        for (uint32_t c = start_c; c < C && num_tiles_read < num_tiles; ++c, start_th = 0) {
+            for (uint32_t th = start_th; th < Ht && num_tiles_read < num_tiles; ++th, start_tw = 0) {
+                cb_wait_front(tt::CBIndex::c_0, 1);
```
Pass CB values as CT args so it's easy to update/bugfix

jvegaTT replied:
Do this in all kernels

jvegaTT commented on ttnn/cpp/ttnn/operations/experimental/bcast_to/device/kernels/compute/compute_interleaved_col_bcast_to.cpp:
```diff
@@ -0,0 +1,56 @@
+    uint32_t N = get_arg_val<uint32_t>(8);
+    uint32_t C = get_arg_val<uint32_t>(9);
+    uint32_t Ht = get_arg_val<uint32_t>(10);
+    uint32_t Wt = get_arg_val<uint32_t>(11);
```
You can use an argument index rather than hardcoding

jvegaTT replied:
Do this in all your kernels

dchenTT replied:
Having an index starting from 0 then increment? OK.

jvegaTT replied:
yes

jvegaTT reviewed (commented):
Added some comments to address

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Address code review. Make cd index as compiler arg, etc.

#16015 Add explicit BroadcastOp for TTNN

Forgot a file.

jvegaTT reviewed (approved):
Good work! This addresses all my comments

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Shorten the profile test time

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Disable bfloat8_b support as it seems unstable. Individual unit tests pass but not CI.

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Fixed scalar bcast kernel hang for certain shpe. Cause: kernel used a wrong variable name, almost like a typo.

dchenTT committed:
#16015 Add explicit BroadcastOp for TTNN

Disable bf8_b support for now as underlying LLK may have un-init() issue.
